comp/core/remoteagent/impl/remoteagent.go

```diff
@@ -0,0 +1,379 @@
+type dependencies struct {
+	Config    config.Component
+	Lifecycle compdef.Lifecycle
+}
+
+// Provides defines the output of the remoteagent component
+type Provides struct {
+	Comp          remoteagent.Component
+	FlareProvider flaretypes.Provider
+	Status        status.InformationProvider
+}
+
+// NewComponent creates a new remoteagent component
+func NewComponent(deps dependencies) Provides {
+	c := newRemoteAgent(deps)
+
+	return Provides{
+		Comp:          c,
+		FlareProvider: flaretypes.NewProvider(c.(*RemoteAgent).fillFlare),
+		Status:        status.NewInformationProvider(remoteagentStatus.GetProvider(c)),
+	}
+}
+
+func newRemoteAgent(deps dependencies) remoteagent.Component {
+	shutdownChan := make(chan struct{})
+	comp := &RemoteAgent{
+		conf:         deps.Config,
+		agentMap:     make(map[string]*remoteAgentDetails),
+		shutdownChan: shutdownChan,
+	}
+
+	deps.Lifecycle.Append(compdef.Hook{
+		OnStart: func(context.Context) error {
+			go comp.start()
+			return nil
+		},
+		OnStop: func(context.Context) error {
+			shutdownChan <- struct{}{}
+			return nil
+		},
+	})
+
+	return comp
+}
+
+// RemoteAgent is the main registry for remote agents. It tracks which remote agents are currently registered, when
+// they were last seen, and handles collecting status and flare data from them on request.
+type RemoteAgent struct {
+	conf         config.Component
+	agentMap     map[string]*remoteAgentDetails
+	agentMapMu   sync.Mutex
+	shutdownChan chan struct{}
+}
+
+// RegisterRemoteAgent registers a remote agent with the registry.
+//
+// If the remote agent is not present in the registry, it is added. If a remote agent with the same ID is already
+// present, the API endpoint and display name are checked: if they are the same, then the "last seen" time of the remote
+// agent is updated, otherwise the remote agent is removed and replaced with the incoming one.
+func (ra *RemoteAgent) RegisterRemoteAgent(registration *remoteagent.RegistrationData) (uint32, error) {
+	recommendedRefreshInterval := uint32(ra.conf.GetDuration("remote_agent.recommended_refresh_interval"))
+
+	ra.agentMapMu.Lock()
+	defer ra.agentMapMu.Unlock()
+
+	// If the remote agent is already registered, then we might be dealing with an update (i.e., periodic check-in) or a
+	// brand new remote agent. As the agent ID may very well not be a unique ID every single time (it could always just
+	// be `process-agent` or what have you), we differentiate between the two scenarios by checking the human name and
+	// API endpoint give to us.
+	//
+	// If either of them are different, then we remove the old remote agent and add the new one. If they're the same,
+	// then we just update the last seen time and move on.
+	agentID := util.SanitizeAgentID(registration.AgentID)
+	entry, ok := ra.agentMap[agentID]
+	if !ok {
+		// We've got a brand new remote agent, so do a sanity check by trying to connect to their gRPC endpoint and if
+		// that works, add them to the registry.
```

Comment:
By looking at the code it does not seems we are trying to connect to the gRPC endpoint as a sanity check. Could you expand on this?

Reply:
No, this was intentional although not for a particularly strong/opinionated reason.

We could definitely have it emulate the old `grpc.DialContext` behavior to check for endpoint validity before returning from the register call. I just went with `NewClient` because it's the recommended default approach, and connection errors are going to get surfaced whether we do it here or as part of trying to query the remote agent when querying for status/flare data.

Do you want me to explicitly have it connect here?

Reply:
> Do you want me to explicitly have it connect here?

Probably just update the comment to reflect the actual behavior 😄